### DISCUSSION TITLE
Creating complex systems?

### DISCUSSION THREAD

**Discussion by cherkasovn** [pymeasure/pymeasure#611] [Q&A]
I am looking for a Python platform to automate chemical experiments.
The key is to control multiple instruments and (most frequently) perform sequences with some interactions between the instruments.

Adding new instruments to PyMeasure is simple and works well. Yet, the challenge is that various instruments have various data rates and some asynchronous communication would be great.

For example, I have several liquid pumps doing something and a measuring device looking at the signals. The pumps need only occasional inputs, while the measurements work frequently. Sending any command to the pump "blocks" the programme and limits usefulness of the measurement data,

Is there any way to do instrument interactions in an asynchronous way?

**Comment by neal-kepler** [comment]
This problem and a consequence of trying to fix it is discussed in #506 . You might find that a good place to start for ideas on how to implement it (either asyncio or in the suggestion by @bilderbuchi of the way it was proposed for the anaheim motion controller). 

Otherwise I defer to others more versed in async programming for more thorough suggestions

**Comment by bilderbuchi** [comment]
Exactly, you could use separate threads or concurrent.futures to separate instrument communication.
E.g. something like this sketch:
```
with ThreadPoolExecutor(max_workers=len(my_instruments)) as executor:
    futures = {}  # future-name coupling for the logging below (optional)
    for instr in my_instruments:
        futures[executor.submit(instr.do_update)] = instr.name  # schedule whatever function call you want to do here
    for fut in as_completed(futures):
        LOGGER.debug(f'Finished comms with {futures[fut]}')
```
This parallelizes communication with all instruments, I was using something similar for fetching updated state. 
If you want to do fully asychronous communication at different rates, you need to adapt this of course. E.g. in the as_completed loop immediately submit a new update for the instrument that just finished. That should work, but there are surely a number of resources online dealing with your specific use case, including proper termination etc.

**Comment by bilderbuchi** [comment]
As this is more of a usage question, I'll convert this into a discussion.

**Comment by BenediktBurger** [comment]
We use another approach in our lab: Most devices have their own control program (sometimes a few devices are controlled by one) and they report their (regularly read or triggered by a callback/new value) measurement data to a central program, which logs the data and offers the possibility to plot any input data vs. any other input data (you can spawn as many plot windows as you desire).
For communication between the programs, I originally used "socket", but inspired by pymeasure, I changed to zmq, which greatly simplifies the process.
Maybe a mixture of multithreading and several independent programs is your way.
